examples/retrieval/retrieval.cpp

```diff
@@ -0,0 +1,358 @@
+#include "common.h"
+#include "llama.h"
+
+#include <algorithm>
+#include <fstream>
+
+struct retrieval_params {
+    std::vector<std::string> context_files;// context files to embed
+    int32_t chunk_size            = 64;    // chunk size for context embedding
+    std::string chunk_separator   = "\n";  // chunk separator for context embedding
+};
+
+static void retrieval_params_print_usage(int argc, char** argv, gpt_params & gpt_params, retrieval_params & params) {
+    fprintf(stderr, "usage: retrieval [options]\n");
+    fprintf(stderr, "options:\n");
+    printf("  --context-files FNAME1 FNAME2...\n");
+    printf("                        files containing context to embed.\n");
+    printf("  --chunk-size N        minimum length of embedded text chunk (default:%d)\n", params.chunk_size);
+    printf("  --chunk-separator STRING\n");
+    printf("                        string to separate chunks (default: \"\\n\")\n");
+    gpt_print_usage(argc, argv, gpt_params);
```

Comment:
Inside `gpt_print_usage` we already have the line `usage: ...`

I think here it's better to print `gpt_print_usage` first, then we print retrieval-related options

Reply:
Yeah this doesn't look that nice ☹️ I'll fix it as proposed.
<img width="578" alt="image" src="https://github.com/ggerganov/llama.cpp/assets/54794500/c85502e4-b1d7-43bb-b856-02d82a23afef">
Although the retrieval parameters(which will likely be used the most) being shown at the bottom of the list does seem inconvenient 🤔

Reply:
Just found out that printing gpt params first opens a new door of problems 😢 
<img width="842" alt="image" src="https://github.com/ggerganov/llama.cpp/assets/54794500/c8be418b-f783-47f8-8706-ab7f31f1b35f">

Reply:
fixed as:
<img width="843" alt="image" src="https://github.com/ggerganov/llama.cpp/assets/54794500/1ec178b2-8a1e-41b5-8e1d-f72a69a23194">